File: src/main/java/kr/syeyoung/dungeonsguide/mod/whosonline/WhosOnlineManager.java
package kr.syeyoung.dungeonsguide.mod.whosonline;

import com.google.common.util.concurrent.ThreadFactoryBuilder;
import com.google.gson.Gson;
import kr.syeyoung.dungeonsguide.mod.whosonline.api.WhosOnlineApi;
import kr.syeyoung.dungeonsguide.mod.whosonline.api.WhosOnlineCache;
import kr.syeyoung.dungeonsguide.mod.whosonline.api.WhosOnlineWebSocket;
import lombok.Getter;
import lombok.val;
import net.minecraft.client.Minecraft;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.fml.common.eventhandler.Event;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.gameevent.TickEvent;
import net.minecraftforge.fml.relauncher.Side;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.concurrent.*;

public class WhosOnlineManager {

    public static final Gson gson = new Gson();

    static final Logger logger = LogManager.getLogger("WhosOnlineManager");

    private final ExecutorService ex;
    private final ScheduledExecutorService se;

    final String remoteHost;
    @Getter
    private WhosOnlineWebSocket websocketManager;
    @Getter
    private WhosOnlineApi whosOnlineApi;
    @Getter
    private WhosOnlineCache cache;

    public WhosOnlineManager(String host) {
        remoteHost = host;
        MinecraftForge.EVENT_BUS.register(this);

        val namedThreadFactory = new ThreadFactoryBuilder().setNameFormat("Dg WhosOnline pool").build();

        ex = Executors.newCachedThreadPool(namedThreadFactory);
        se = Executors.newScheduledThreadPool(2, namedThreadFactory);
    }

    public void init() {
        this.cache = new WhosOnlineCache();
        this.websocketManager = new WhosOnlineWebSocket(remoteHost, se, cache, Minecraft.getMinecraft().getSession().getPlayerID());
        websocketManager.connect();
        this.whosOnlineApi = new WhosOnlineApi(websocketManager, cache,ex);

    }

    boolean closed = false;

    public void close(){
        closed = true;
        try {
            this.websocketManager.close();
            this.websocketManager = null;
            this.whosOnlineApi = null;
            ex.awaitTermination(1,TimeUnit.SECONDS);
            se.awaitTermination(1,TimeUnit.SECONDS);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @SubscribeEvent
    public void websocketdied(WhosOnlineDied e){
        if(closed) return;
        logger.info("Who'sOnline websocket died, trying again in 4 seconds");
        se.schedule(this::init, 4, TimeUnit.SECONDS);
    }

    public static class WhosOnlineDied extends Event {

    }

}

File: src/main/java/kr/syeyoung/dungeonsguide/whosonline/WhosOnlineManager.kt
package kr.syeyoung.dungeonsguide.whosonline

import com.google.common.util.concurrent.ThreadFactoryBuilder
import com.google.gson.Gson
import kr.syeyoung.dungeonsguide.whosonline.api.WhosOnlineApi
import kr.syeyoung.dungeonsguide.whosonline.api.WhosOnlineCache
import kr.syeyoung.dungeonsguide.whosonline.api.WhosOnlineRest
import kr.syeyoung.dungeonsguide.whosonline.api.WhosOnlineWebSocket
import lombok.Getter
import net.minecraft.client.Minecraft
import net.minecraftforge.common.MinecraftForge
import net.minecraftforge.fml.common.eventhandler.Event
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import org.apache.logging.log4j.LogManager
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit

class WhosOnlineManager(val remoteHost: String) {
    private val ex: ExecutorService
    private val se: ScheduledExecutorService

    fun getWebsocketClient(): WhosOnlineApi? {
        return websocketClient
    }

    private var webSocket: WhosOnlineWebSocket? = null

    @Getter
    private var websocketClient: WhosOnlineApi? = null

    @Getter
    private var cache: WhosOnlineCache? = null
    private var restClient: WhosOnlineRest? = null
    var useDebugServers = false
    fun init() {
        cache = WhosOnlineCache()
        val websocketUri = if (useDebugServers) "ws://$remoteHost/ws" else "wss://$remoteHost/ws"
        val restUri = if (useDebugServers) "http://$remoteHost" else "https://$remoteHost"
        webSocket = WhosOnlineWebSocket(websocketUri, se, cache, Minecraft.getMinecraft().session.playerID)
        restClient = WhosOnlineRest(cache, ex, restUri)
        webSocket!!.connect()
        websocketClient = WhosOnlineApi(webSocket!!, cache!!, ex)

    }

    var closed = false

    init {
        if (remoteHost.startsWith("localhost")) {
            useDebugServers = true
        }
        MinecraftForge.EVENT_BUS.register(this)
        val namedThreadFactory = ThreadFactoryBuilder().setNameFormat("Dg WhosOnline pool").build()
        ex = Executors.newCachedThreadPool(namedThreadFactory)
        se = Executors.newScheduledThreadPool(2, namedThreadFactory)
    }

    fun close() {
        closed = true
        try {
            webSocket?.close()
            webSocket = null
            websocketClient = null
            restClient = null
            ex.awaitTermination(1, TimeUnit.SECONDS)
            se.awaitTermination(1, TimeUnit.SECONDS)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    @SubscribeEvent
    fun websocketdied(e: WhosOnlineDied?) {
        if (closed) return
        logger.info("Who'sOnline websocket died, trying again in 4 seconds")
        se.schedule({ init() }, 4, TimeUnit.SECONDS)
    }

    class WhosOnlineDied : Event()
    companion object {
        @JvmField
        val gson = Gson()
        val logger = LogManager.getLogger("WhosOnlineManager")
    }
}